File: web_dashboard/src/lib/assessmentTemplates.ts
export type SelfAssessmentFieldType = 'text' | 'number';

export interface SelfAssessmentField {
  key: string;
  label: string;
  placeholder: string;
  inputType: SelfAssessmentFieldType;
}

export interface SupervisorCriterion {
  key: string;
  label: string;
}

export interface AssessmentTemplateSnapshot {
  selfAssessmentFields: SelfAssessmentField[];
  supervisorCriteria: SupervisorCriterion[];
}

export interface TeacherAssessmentTemplateOverrides {
  hiddenSelfAssessmentFieldKeys: string[];
  hiddenSupervisorCriteriaKeys: string[];
  selfAssessmentOrderKeys: string[];
  supervisorCriteriaOrderKeys: string[];
  additionalSelfAssessmentFields: SelfAssessmentField[];
  additionalSupervisorCriteria: SupervisorCriterion[];
}

export const defaultAssessmentTemplateSnapshot: AssessmentTemplateSnapshot = {
  selfAssessmentFields: [
    {
      key: 'whatDidYouDo',
      label: 'Vad har du fått göra?',
      placeholder: 'Beskriv de arbetsuppgifter du utförde...',
      inputType: 'text',
    },
    {
      key: 'whatWasPositive',
      label: 'Vad har varit positivt med APLen?',
      placeholder: 'Vad har varit bra? Vad har du lärt dig?',
      inputType: 'text',
    },
    {
      key: 'whatCouldBeBetter',
      label: 'Vad skulle kunnat vara bättre?',
      placeholder: 'Vad var utmanande? Vad skulle kunna förbättras?',
      inputType: 'text',
    },
    {
      key: 'whatCouldYouDoDifferently',
      label: 'Vad kunde du som elev gjort annorlunda?',
      placeholder: 'Hur kunde du bidragit mer? Vad kan du förbättra till nästa gång?',
      inputType: 'text',
    },
    {
      key: 'overallRating',
      label: 'Vilket betyg för din APL-period? (1-10)',
      placeholder: '1=mindre bra, 10=fantastiskt',
      inputType: 'number',
    },
  ],
  supervisorCriteria: [
    { key: 'engagement', label: 'Engagemang' },
    { key: 'initiative', label: 'Initiativtagande' },
    { key: 'collaboration', label: 'Samarbetsförmåga' },
    { key: 'problemSolving', label: 'Problemlösning' },
    { key: 'workQuality', label: 'Kvalitet på arbete' },
  ],
};

function sanitizeKeyPart(value: string): string {
  return value
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '');
}

function matchesTemplateKey(first: string, second: string): boolean {
  const normalizedFirst = first.trim();
  const normalizedSecond = second.trim();

  if (!normalizedFirst || !normalizedSecond) return false;

  return (
    normalizedFirst === normalizedSecond ||
    sanitizeKeyPart(normalizedFirst) === sanitizeKeyPart(normalizedSecond)
  );
}

function ensureUniqueKey(baseKey: string, usedKeys: Set<string>): string {
  let nextKey = baseKey || 'field';
  let suffix = 2;
  while (usedKeys.has(nextKey)) {
    nextKey = `${baseKey || 'field'}_${suffix}`;
    suffix += 1;
  }
  usedKeys.add(nextKey);
  return nextKey;
}

export function createAssessmentFieldKey(label: string, existingKeys: string[]): string {
  const usedKeys = new Set(existingKeys);
  const baseKey = sanitizeKeyPart(label) || 'field';
  return ensureUniqueKey(baseKey, usedKeys);
}

function isObjectRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object';
}

function sanitizeSelfAssessmentFields(rawFields: unknown): SelfAssessmentField[] {
  const usedSelfKeys = new Set<string>();

  return Array.isArray(rawFields)
    ? rawFields
        .filter(isObjectRecord)
        .map((field) => {
          const label = String(field.label ?? '').trim();
          if (!label) return null;

          const requestedKey = sanitizeKeyPart(String(field.key ?? '').trim()) || sanitizeKeyPart(label);
          const key = ensureUniqueKey(requestedKey || 'field', usedSelfKeys);
          const inputType = field.inputType === 'number' ? 'number' : 'text';

          return {
            key,
            label,
            placeholder: String(field.placeholder ?? '').trim(),
            inputType,
          } satisfies SelfAssessmentField;
        })
        .filter((field): field is SelfAssessmentField => field !== null)
    : [];
}

function sanitizeSupervisorCriteria(rawCriteria: unknown): SupervisorCriterion[] {
  const usedCriteriaKeys = new Set<string>();

  return Array.isArray(rawCriteria)
    ? rawCriteria
        .filter(isObjectRecord)
        .map((criterion) => {
          const label = String(criterion.label ?? '').trim();
          if (!label) return null;

          const requestedKey = sanitizeKeyPart(String(criterion.key ?? '').trim()) || sanitizeKeyPart(label);
          const key = ensureUniqueKey(requestedKey || 'criterion', usedCriteriaKeys);

          return {
            key,
            label,
          } satisfies SupervisorCriterion;
        })
        .filter((criterion): criterion is SupervisorCriterion => criterion !== null)
    : [];
}

function sanitizeOrderKeys(rawKeys: unknown): string[] {
  return Array.isArray(rawKeys)
    ? Array.from(
        new Set(
          rawKeys
            .map((key) => String(key ?? '').trim())
            .filter(Boolean),
        ),
      )
    : [];
}

function orderItemsByKeys<T extends { key: string }>(items: T[], orderKeys: string[]): T[] {
  if (items.length <= 1 || orderKeys.length === 0) return [...items];

  const remaining = [...items];
  const ordered: T[] = [];

  for (const orderKey of orderKeys) {
    const index = remaining.findIndex((item) => matchesTemplateKey(orderKey, item.key));
    if (index === -1) continue;
    ordered.push(remaining.splice(index, 1)[0]);
  }

  ordered.push(...remaining);
  return ordered;
}

export function orderTeacherSelfAssessmentFieldsForView(
  baseSnapshot: AssessmentTemplateSnapshot,
  overrides: TeacherAssessmentTemplateOverrides,
): SelfAssessmentField[] {
  return orderItemsByKeys(
    [
      ...baseSnapshot.selfAssessmentFields,
      ...overrides.additionalSelfAssessmentFields,
    ],
    overrides.selfAssessmentOrderKeys,
  );
}

export function orderTeacherSupervisorCriteriaForView(
  baseSnapshot: AssessmentTemplateSnapshot,
  overrides: TeacherAssessmentTemplateOverrides,
): SupervisorCriterion[] {
  return orderItemsByKeys(
    [
      ...baseSnapshot.supervisorCriteria,
      ...overrides.additionalSupervisorCriteria,
    ],
    overrides.supervisorCriteriaOrderKeys,
  );
}

export function sanitizeAssessmentTemplateSnapshot(
  raw: Partial<AssessmentTemplateSnapshot> | null | undefined,
): AssessmentTemplateSnapshot {
  const selfAssessmentFields = sanitizeSelfAssessmentFields(raw?.selfAssessmentFields);
  const supervisorCriteria = sanitizeSupervisorCriteria(raw?.supervisorCriteria);

  return {
    selfAssessmentFields:
      selfAssessmentFields.length > 0
        ? selfAssessmentFields
        : defaultAssessmentTemplateSnapshot.selfAssessmentFields,
    supervisorCriteria:
      supervisorCriteria.length > 0
        ? supervisorCriteria
        : defaultAssessmentTemplateSnapshot.supervisorCriteria,
  };
}

export function sanitizeTeacherAssessmentTemplateOverrides(
  raw: Partial<TeacherAssessmentTemplateOverrides> | null | undefined,
): TeacherAssessmentTemplateOverrides {
  return {
    hiddenSelfAssessmentFieldKeys: sanitizeOrderKeys(raw?.hiddenSelfAssessmentFieldKeys),
    hiddenSupervisorCriteriaKeys: sanitizeOrderKeys(raw?.hiddenSupervisorCriteriaKeys),
    selfAssessmentOrderKeys: sanitizeOrderKeys(raw?.selfAssessmentOrderKeys),
    supervisorCriteriaOrderKeys: sanitizeOrderKeys(raw?.supervisorCriteriaOrderKeys),
    additionalSelfAssessmentFields: sanitizeSelfAssessmentFields(raw?.additionalSelfAssessmentFields),
    additionalSupervisorCriteria: sanitizeSupervisorCriteria(raw?.additionalSupervisorCriteria),
  };
}

export function mergeAssessmentTemplateSnapshot(
  baseSnapshot: AssessmentTemplateSnapshot,
  overrides: TeacherAssessmentTemplateOverrides,
): AssessmentTemplateSnapshot {
  const hiddenSelfKeys = overrides.hiddenSelfAssessmentFieldKeys.map((key) => key.trim()).filter(Boolean);
  const hiddenSupervisorKeys = overrides.hiddenSupervisorCriteriaKeys.map((key) => key.trim()).filter(Boolean);

  const selfAssessmentFields: SelfAssessmentField[] = [];
  const usedSelfKeys = new Set<string>();

  for (const field of baseSnapshot.selfAssessmentFields) {
    if (hiddenSelfKeys.some((key) => matchesTemplateKey(key, field.key))) continue;
    selfAssessmentFields.push(field);
    usedSelfKeys.add(field.key);
  }

  for (const field of overrides.additionalSelfAssessmentFields) {
    const key = ensureUniqueKey(sanitizeKeyPart(field.key) || 'field', usedSelfKeys);
    selfAssessmentFields.push({
      key,
      label: field.label,
      placeholder: field.placeholder,
      inputType: field.inputType,
    });
  }

  const orderedSelfAssessmentFields = orderItemsByKeys(
    selfAssessmentFields,
    overrides.selfAssessmentOrderKeys,
  );

  const supervisorCriteria: SupervisorCriterion[] = [];
  const usedCriteriaKeys = new Set<string>();

  for (const criterion of baseSnapshot.supervisorCriteria) {
    if (hiddenSupervisorKeys.some((key) => matchesTemplateKey(key, criterion.key))) continue;
    supervisorCriteria.push(criterion);
    usedCriteriaKeys.add(criterion.key);
  }

  for (const criterion of overrides.additionalSupervisorCriteria) {
    const key = ensureUniqueKey(sanitizeKeyPart(criterion.key) || 'criterion', usedCriteriaKeys);
    supervisorCriteria.push({ key, label: criterion.label });
  }

  const orderedSupervisorCriteria = orderItemsByKeys(
    supervisorCriteria,
    overrides.supervisorCriteriaOrderKeys,
  );

  return {
    selfAssessmentFields:
      orderedSelfAssessmentFields.length > 0
        ? orderedSelfAssessmentFields
        : baseSnapshot.selfAssessmentFields,
    supervisorCriteria:
      orderedSupervisorCriteria.length > 0
        ? orderedSupervisorCriteria
        : baseSnapshot.supervisorCriteria,
  };
}

export function hasTeacherAssessmentOverrides(overrides: TeacherAssessmentTemplateOverrides): boolean {
  return (
    overrides.hiddenSelfAssessmentFieldKeys.length > 0 ||
    overrides.hiddenSupervisorCriteriaKeys.length > 0 ||
    overrides.selfAssessmentOrderKeys.length > 0 ||
    overrides.supervisorCriteriaOrderKeys.length > 0 ||
    overrides.additionalSelfAssessmentFields.length > 0 ||
    overrides.additionalSupervisorCriteria.length > 0
  );
}

export function getSelfAssessmentLabel(
  snapshot: AssessmentTemplateSnapshot | null | undefined,
  key: string,
): string {
  return (
    snapshot?.selfAssessmentFields.find((field) => field.key === key)?.label ??
    defaultAssessmentTemplateSnapshot.selfAssessmentFields.find((field) => field.key === key)?.label ??
    key
  );
}

export function getAssessmentCriterionLabel(
  snapshot: AssessmentTemplateSnapshot | null | undefined,
  key: string,
  value?: { label?: string } | null,
): string {
  return (
    value?.label?.trim() ||
    snapshot?.supervisorCriteria.find((criterion) => criterion.key === key)?.label ||
    defaultAssessmentTemplateSnapshot.supervisorCriteria.find((criterion) => criterion.key === key)?.label ||
    key
  );
}

File: lib/core/assessment_templates.dart
import 'package:cloud_firestore/cloud_firestore.dart';

const _teacherAssessmentTemplatesCollection = 'teacherAssessmentTemplates';

class SelfAssessmentField {
  final String key;
  final String label;
  final String placeholder;
  final String inputType;

  const SelfAssessmentField({
    required this.key,
    required this.label,
    this.placeholder = '',
    this.inputType = 'text',
  });

  Map<String, dynamic> toJson() => {
    'key': key,
    'label': label,
    'placeholder': placeholder,
    'inputType': inputType,
  };
}

class SupervisorCriterion {
  final String key;
  final String label;

  const SupervisorCriterion({required this.key, required this.label});

  Map<String, dynamic> toJson() => {'key': key, 'label': label};
}

class AssessmentTemplateConfig {
  final List<SelfAssessmentField> selfAssessmentFields;
  final List<SupervisorCriterion> supervisorCriteria;

  const AssessmentTemplateConfig({
    required this.selfAssessmentFields,
    required this.supervisorCriteria,
  });

  Map<String, dynamic> toJson() => {
    'selfAssessmentFields': selfAssessmentFields
        .map((field) => field.toJson())
        .toList(),
    'supervisorCriteria': supervisorCriteria
        .map((criterion) => criterion.toJson())
        .toList(),
  };
}

class TeacherAssessmentTemplateOverrides {
  final List<String> hiddenSelfAssessmentFieldKeys;
  final List<String> hiddenSupervisorCriteriaKeys;
  final List<String> selfAssessmentOrderKeys;
  final List<String> supervisorCriteriaOrderKeys;
  final List<SelfAssessmentField> additionalSelfAssessmentFields;
  final List<SupervisorCriterion> additionalSupervisorCriteria;

  const TeacherAssessmentTemplateOverrides({
    this.hiddenSelfAssessmentFieldKeys = const [],
    this.hiddenSupervisorCriteriaKeys = const [],
    this.selfAssessmentOrderKeys = const [],
    this.supervisorCriteriaOrderKeys = const [],
    this.additionalSelfAssessmentFields = const [],
    this.additionalSupervisorCriteria = const [],
  });

  bool get isEmpty =>
      hiddenSelfAssessmentFieldKeys.isEmpty &&
      hiddenSupervisorCriteriaKeys.isEmpty &&
      selfAssessmentOrderKeys.isEmpty &&
      supervisorCriteriaOrderKeys.isEmpty &&
      additionalSelfAssessmentFields.isEmpty &&
      additionalSupervisorCriteria.isEmpty;

  Map<String, dynamic> toJson() => {
    'hiddenSelfAssessmentFieldKeys': hiddenSelfAssessmentFieldKeys,
    'hiddenSupervisorCriteriaKeys': hiddenSupervisorCriteriaKeys,
    'selfAssessmentOrderKeys': selfAssessmentOrderKeys,
    'supervisorCriteriaOrderKeys': supervisorCriteriaOrderKeys,
    'additionalSelfAssessmentFields': additionalSelfAssessmentFields
        .map((field) => field.toJson())
        .toList(),
    'additionalSupervisorCriteria': additionalSupervisorCriteria
        .map((criterion) => criterion.toJson())
        .toList(),
  };
}

const defaultAssessmentTemplateConfig = AssessmentTemplateConfig(
  selfAssessmentFields: [
    SelfAssessmentField(
      key: 'whatDidYouDo',
      label: 'Vad har du fått göra?',
      placeholder: 'Beskriv de arbetsuppgifter du utförde...',
    ),
    SelfAssessmentField(
      key: 'whatWasPositive',
      label: 'Vad har varit positivt med APLen?',
      placeholder: 'Vad har varit bra? Vad har du lärt dig?',
    ),
    SelfAssessmentField(
      key: 'whatCouldBeBetter',
      label: 'Vad skulle kunnat vara bättre?',
      placeholder: 'Vad var utmanande? Vad skulle kunna förbättras?',
    ),
    SelfAssessmentField(
      key: 'whatCouldYouDoDifferently',
      label: 'Vad kunde du som elev gjort annorlunda?',
      placeholder:
          'Hur kunde du bidragit mer? Vad kan du förbättra till nästa gång?',
    ),
    SelfAssessmentField(
      key: 'overallRating',
      label: 'Vilket betyg för din APL-period? (1-10)',
      placeholder: '1=mindre bra, 10=fantastiskt',
      inputType: 'number',
    ),
  ],
  supervisorCriteria: [
    SupervisorCriterion(key: 'engagement', label: 'Engagemang'),
    SupervisorCriterion(key: 'initiative', label: 'Initiativtagande'),
    SupervisorCriterion(key: 'collaboration', label: 'Samarbetsförmåga'),
    SupervisorCriterion(key: 'problemSolving', label: 'Problemlösning'),
    SupervisorCriterion(key: 'workQuality', label: 'Kvalitet på arbete'),
  ],
);

String _sanitizeKeyPart(String value) {
  final normalized = value
      .toLowerCase()
      .replaceAll(RegExp(r'[åä]'), 'a')
      .replaceAll('ö', 'o')
      .replaceAll(RegExp(r'[^a-z0-9]+'), '_')
      .replaceAll(RegExp(r'^_+|_+$'), '');
  return normalized;
}

bool _matchesTemplateKey(String first, String second) {
  final normalizedFirst = first.trim();
  final normalizedSecond = second.trim();
  if (normalizedFirst.isEmpty || normalizedSecond.isEmpty) {
    return false;
  }

  return normalizedFirst == normalizedSecond ||
      _sanitizeKeyPart(normalizedFirst) == _sanitizeKeyPart(normalizedSecond);
}

String _ensureUniqueKey(String baseKey, Set<String> usedKeys, String fallback) {
  var key = baseKey.isNotEmpty ? baseKey : fallback;
  var suffix = 2;
  while (usedKeys.contains(key)) {
    key = '${baseKey.isNotEmpty ? baseKey : fallback}_$suffix';
    suffix += 1;
  }
  usedKeys.add(key);
  return key;
}

List<SelfAssessmentField> _sanitizeSelfAssessmentFields(dynamic rawFields) {
  final usedSelfKeys = <String>{};
  final selfAssessmentFields = <SelfAssessmentField>[];

  if (rawFields is List) {
    for (final entry in rawFields.whereType<Map>()) {
      final label = (entry['label'] ?? '').toString().trim();
      if (label.isEmpty) continue;

      final requestedKey = _sanitizeKeyPart(
        (entry['key'] ?? '').toString().trim(),
      );
      final key = _ensureUniqueKey(
        requestedKey.isNotEmpty ? requestedKey : _sanitizeKeyPart(label),
        usedSelfKeys,
        'field',
      );

      selfAssessmentFields.add(
        SelfAssessmentField(
          key: key,
          label: label,
          placeholder: (entry['placeholder'] ?? '').toString().trim(),
          inputType: (entry['inputType'] ?? 'text').toString() == 'number'
              ? 'number'
              : 'text',
        ),
      );
    }
  }

  return selfAssessmentFields;
}

List<SupervisorCriterion> _sanitizeSupervisorCriteria(dynamic rawCriteria) {
  final usedCriteriaKeys = <String>{};
  final supervisorCriteria = <SupervisorCriterion>[];

  if (rawCriteria is List) {
    for (final entry in rawCriteria.whereType<Map>()) {
      final label = (entry['label'] ?? '').toString().trim();
      if (label.isEmpty) continue;

      final requestedKey = _sanitizeKeyPart(
        (entry['key'] ?? '').toString().trim(),
      );
      final key = _ensureUniqueKey(
        requestedKey.isNotEmpty ? requestedKey : _sanitizeKeyPart(label),
        usedCriteriaKeys,
        'criterion',
      );

      supervisorCriteria.add(SupervisorCriterion(key: key, label: label));
    }
  }

  return supervisorCriteria;
}

List<String> _sanitizeOrderKeys(dynamic rawKeys) {
  if (rawKeys is! List) return const <String>[];

  return rawKeys
      .map((value) => value.toString().trim())
      .where((value) => value.isNotEmpty)
      .toSet()
      .toList();
}

List<SelfAssessmentField> _orderSelfAssessmentFields(
  List<SelfAssessmentField> items,
  List<String> orderKeys,
) {
  if (items.length <= 1 || orderKeys.isEmpty) {
    return List<SelfAssessmentField>.from(items);
  }

  final remaining = List<SelfAssessmentField>.from(items);
  final ordered = <SelfAssessmentField>[];

  for (final orderKey in orderKeys) {
    final index = remaining.indexWhere(
      (field) => _matchesTemplateKey(orderKey, field.key),
    );
    if (index == -1) continue;
    ordered.add(remaining.removeAt(index));
  }

  ordered.addAll(remaining);
  return ordered;
}

List<SupervisorCriterion> _orderSupervisorCriteria(
  List<SupervisorCriterion> items,
  List<String> orderKeys,
) {
  if (items.length <= 1 || orderKeys.isEmpty) {
    return List<SupervisorCriterion>.from(items);
  }

  final remaining = List<SupervisorCriterion>.from(items);
  final ordered = <SupervisorCriterion>[];

  for (final orderKey in orderKeys) {
    final index = remaining.indexWhere(
      (criterion) => _matchesTemplateKey(orderKey, criterion.key),
    );
    if (index == -1) continue;
    ordered.add(remaining.removeAt(index));
  }

  ordered.addAll(remaining);
  return ordered;
}

AssessmentTemplateConfig sanitizeAssessmentTemplateConfig(dynamic raw) {
  final selfAssessmentFields = _sanitizeSelfAssessmentFields(
    raw is Map ? raw['selfAssessmentFields'] : null,
  );
  final supervisorCriteria = _sanitizeSupervisorCriteria(
    raw is Map ? raw['supervisorCriteria'] : null,
  );

  return AssessmentTemplateConfig(
    selfAssessmentFields: selfAssessmentFields.isNotEmpty
        ? selfAssessmentFields
        : defaultAssessmentTemplateConfig.selfAssessmentFields,
    supervisorCriteria: supervisorCriteria.isNotEmpty
        ? supervisorCriteria
        : defaultAssessmentTemplateConfig.supervisorCriteria,
  );
}

TeacherAssessmentTemplateOverrides sanitizeTeacherAssessmentTemplateOverrides(
  dynamic raw,
) {
  final hiddenSelfAssessmentFieldKeys =
      raw is Map && raw['hiddenSelfAssessmentFieldKeys'] is List
      ? (raw['hiddenSelfAssessmentFieldKeys'] as List)
            .map((value) => value.toString().trim())
            .where((value) => value.isNotEmpty)
            .toSet()
            .toList()
      : <String>[];

  final hiddenSupervisorCriteriaKeys =
      raw is Map && raw['hiddenSupervisorCriteriaKeys'] is List
      ? (raw['hiddenSupervisorCriteriaKeys'] as List)
            .map((value) => value.toString().trim())
            .where((value) => value.isNotEmpty)
            .toSet()
            .toList()
      : <String>[];

  return TeacherAssessmentTemplateOverrides(
    hiddenSelfAssessmentFieldKeys: hiddenSelfAssessmentFieldKeys,
    hiddenSupervisorCriteriaKeys: hiddenSupervisorCriteriaKeys,
    selfAssessmentOrderKeys: _sanitizeOrderKeys(
      raw is Map ? raw['selfAssessmentOrderKeys'] : null,
    ),
    supervisorCriteriaOrderKeys: _sanitizeOrderKeys(
      raw is Map ? raw['supervisorCriteriaOrderKeys'] : null,
    ),
    additionalSelfAssessmentFields: _sanitizeSelfAssessmentFields(
      raw is Map ? raw['additionalSelfAssessmentFields'] : null,
    ),
    additionalSupervisorCriteria: _sanitizeSupervisorCriteria(
      raw is Map ? raw['additionalSupervisorCriteria'] : null,
    ),
  );
}

AssessmentTemplateConfig mergeAssessmentTemplateConfig(
  AssessmentTemplateConfig baseConfig,
  TeacherAssessmentTemplateOverrides overrides,
) {
  final hiddenSelfKeys = overrides.hiddenSelfAssessmentFieldKeys
      .map((key) => key.trim())
      .where((key) => key.isNotEmpty)
      .toList();
  final hiddenSupervisorKeys = overrides.hiddenSupervisorCriteriaKeys
      .map((key) => key.trim())
      .where((key) => key.isNotEmpty)
      .toList();

  final selfAssessmentFields = <SelfAssessmentField>[];
  final usedSelfKeys = <String>{};

  for (final field in baseConfig.selfAssessmentFields) {
    if (hiddenSelfKeys.any((key) => _matchesTemplateKey(key, field.key))) {
      continue;
    }
    selfAssessmentFields.add(field);
    usedSelfKeys.add(field.key);
  }

  for (final field in overrides.additionalSelfAssessmentFields) {
    final mergedKey = _ensureUniqueKey(
      _sanitizeKeyPart(field.key),
      usedSelfKeys,
      'field',
    );
    selfAssessmentFields.add(
      SelfAssessmentField(
        key: mergedKey,
        label: field.label,
        placeholder: field.placeholder,
        inputType: field.inputType,
      ),
    );
  }

  final orderedSelfAssessmentFields = _orderSelfAssessmentFields(
    selfAssessmentFields,
    overrides.selfAssessmentOrderKeys,
  );

  final supervisorCriteria = <SupervisorCriterion>[];
  final usedSupervisorKeys = <String>{};

  for (final criterion in baseConfig.supervisorCriteria) {
    if (hiddenSupervisorKeys.any(
      (key) => _matchesTemplateKey(key, criterion.key),
    )) {
      continue;
    }
    supervisorCriteria.add(criterion);
    usedSupervisorKeys.add(criterion.key);
  }

  for (final criterion in overrides.additionalSupervisorCriteria) {
    final mergedKey = _ensureUniqueKey(
      _sanitizeKeyPart(criterion.key),
      usedSupervisorKeys,
      'criterion',
    );
    supervisorCriteria.add(
      SupervisorCriterion(key: mergedKey, label: criterion.label),
    );
  }

  final orderedSupervisorCriteria = _orderSupervisorCriteria(
    supervisorCriteria,
    overrides.supervisorCriteriaOrderKeys,
  );

  return AssessmentTemplateConfig(
    selfAssessmentFields: orderedSelfAssessmentFields.isNotEmpty
        ? orderedSelfAssessmentFields
        : baseConfig.selfAssessmentFields,
    supervisorCriteria: orderedSupervisorCriteria.isNotEmpty
        ? orderedSupervisorCriteria
        : baseConfig.supervisorCriteria,
  );
}

Future<AssessmentTemplateConfig> loadAssessmentTemplateConfig() async {
  try {
    final doc = await FirebaseFirestore.instance
        .collection('appSettings')
        .doc('assessmentTemplates')
        .get();

    if (!doc.exists) return defaultAssessmentTemplateConfig;

    return sanitizeAssessmentTemplateConfig(doc.data());
  } catch (_) {
    return defaultAssessmentTemplateConfig;
  }
}

Future<TeacherAssessmentTemplateOverrides>
loadTeacherAssessmentTemplateOverrides({required String teacherUid}) async {
  if (teacherUid.trim().isEmpty) {
    return const TeacherAssessmentTemplateOverrides();
  }

  try {
    final doc = await FirebaseFirestore.instance
        .collection(_teacherAssessmentTemplatesCollection)
        .doc(teacherUid)
        .get();

    if (!doc.exists) {
      return const TeacherAssessmentTemplateOverrides();
    }

    return sanitizeTeacherAssessmentTemplateOverrides(doc.data());
  } catch (_) {
    return const TeacherAssessmentTemplateOverrides();
  }
}

Future<AssessmentTemplateConfig> loadMergedAssessmentTemplateConfig({
  required String teacherUid,
}) async {
  final baseConfig = await loadAssessmentTemplateConfig();
  if (teacherUid.trim().isEmpty) return baseConfig;

  final overrides = await loadTeacherAssessmentTemplateOverrides(
    teacherUid: teacherUid,
  );
  return mergeAssessmentTemplateConfig(baseConfig, overrides);
}

Future<AssessmentTemplateConfig> loadAssessmentTemplateConfigForStudent(
  String studentUid,
) async {
  try {
    final userDoc = await FirebaseFirestore.instance
        .collection('users')
        .doc(studentUid)
        .get();
    final teacherUid = (userDoc.data()?['teacherUid'] ?? '').toString().trim();
    return loadMergedAssessmentTemplateConfig(teacherUid: teacherUid);
  } catch (_) {
    return loadAssessmentTemplateConfig();
  }
}

String getSelfAssessmentLabel(AssessmentTemplateConfig? config, String key) {
  final source = config ?? defaultAssessmentTemplateConfig;
  for (final field in source.selfAssessmentFields) {
    if (field.key == key) return field.label;
  }
  for (final field in defaultAssessmentTemplateConfig.selfAssessmentFields) {
    if (field.key == key) return field.label;
  }
  return key;
}

String getAssessmentCriterionLabel(
  AssessmentTemplateConfig? config,
  String key,
  Map<dynamic, dynamic>? value,
) {
  final explicitLabel = (value?['label'] ?? '').toString().trim();
  if (explicitLabel.isNotEmpty) return explicitLabel;

  final source = config ?? defaultAssessmentTemplateConfig;
  for (final criterion in source.supervisorCriteria) {
    if (criterion.key == key) return criterion.label;
  }
  for (final criterion in defaultAssessmentTemplateConfig.supervisorCriteria) {
    if (criterion.key == key) return criterion.label;
  }
  return key;
}
